src/Faithlife.Utility/EnumerableUtility.cs

```diff
@@ -765,7 +808,129 @@ static IEnumerable<T> DoWhereNotNull(IEnumerable<T?> s)
+		/// <param name="resultSelector">A function that specifies how to merge the elements from the input sequences.</param>
+		/// <returns>A sequence containing merged elements of the input sequences. If the sequences don't have the same number of items, it stops at the end of the shortest sequence.</returns>
+		public static IEnumerable<TResult> ZipTruncate<T1, T2, T3, TResult>(this IEnumerable<T1> first,
+			IEnumerable<T2> second,
+			IEnumerable<T3> third,
+			Func<T1, T2, T3, TResult> resultSelector) =>
+			ZipImpl(
+				first ?? throw new ArgumentNullException(nameof(first)),
+				second ?? throw new ArgumentNullException(nameof(second)),
+				third ?? throw new ArgumentNullException(nameof(third)),
+				resultSelector ?? throw new ArgumentNullException(nameof(resultSelector)),
+				UnbalancedZipStrategy.Truncate);
+
+		/// <summary>
+		/// Separates a sequence of tuples into a tuple of sequences.
+		/// </summary>
+		/// <param name="source">The sequence to separate.</param>
+		/// <returns>A tuple of sequences, separating the source sequence's <c>Item1</c> elements from its <c>Item2</c> elements.</returns>
+		[SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1414:Tuple types in signatures should have element names", Justification = "By design.")]
+		public static (IEnumerable<T1>, IEnumerable<T2>) Unzip<T1, T2>(this IEnumerable<(T1, T2)> source) =>
+			Unzip(source ?? throw new ArgumentNullException(nameof(source)), x => x.Item1, x => x.Item2);
+
+		/// <summary>
+		/// Separates a sequence of tuples into a tuple of sequences.
+		/// </summary>
+		/// <param name="source">The sequence to separate.</param>
+		/// <returns>A tuple of sequences, separating the source sequence's <c>Item1</c>, <c>Item2</c>, and <c>Item3</c> elements.</returns>
+		[SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1414:Tuple types in signatures should have element names", Justification = "By design.")]
+		public static (IEnumerable<T1>, IEnumerable<T2>, IEnumerable<T3>) Unzip<T1, T2, T3>(this IEnumerable<(T1, T2, T3)> source) =>
+			Unzip(source ?? throw new ArgumentNullException(nameof(source)), x => x.Item1, x => x.Item2, x => x.Item3);
+
+		/// <summary>
+		/// Separates a sequence into a tuple of sequences.
+		/// </summary>
+		/// <param name="source">The sequence to separate.</param>
+		/// <param name="firstSelector">A function for selecting the first return sequence's elements from the source sequence's elements.</param>
+		/// <param name="secondSelector">A function for selecting the second return sequence's elements from the source sequence's elements.</param>
+		/// <returns>A tuple of sequences, separating the source sequence according to the selector arguments.</returns>
+		[SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1414:Tuple types in signatures should have element names", Justification = "By design.")]
+		public static (IEnumerable<T1>, IEnumerable<T2>) Unzip<TSource, T1, T2>(
```

Comment:
Although I can imagine a "lazy" implementation that returns a tuple of two objects that only enumerate the `source` sequence as _they_ are enumerated, it seems unlikely that we'd build that. I'm wondering whether it would just be worth returning `(IReadOnlyList<T1>, IReadOnlyList<T2>)` since the sequences will be materialised.

Reply:
Thinking about this more, the lazy implementation would just be `Unzip(source, firstSelector, secondSelector) => (source.Select(firstSelector), source.Select(secondSelector)`; I'm wondering if that should just be the implementation, since it adds the benefit of laziness (and just depending on framework code).

Reply:
That does enumerate `source` twice, which may be undesirable.

Reply:
Overriding my earlier comments: Let's keep the signature (to allow for future laziness) and the current implementation (which enumerates `source` once).

Reply:
Now that you mention it, that gives me an idea: what if we enumerate the source once at the beginning of the method, then return the `.Select` results as you mentioned? Then half of the process can be lazy, at least.

```c#
var enumeratedSource = source.ToArray();
return (enumeratedSource.Select(firstSelector), enumeratedSource.Select(secondSelector));
```

I've made that change in [`a1bfb97`](https://github.com/Faithlife/FaithlifeUtility/pull/48/commits/a1bfb9799ceb903d7a8cef2c2234c2f917b98be8)—let me know if that looks good. 👍🏼

In that case, would it be good to make the `source` parameter an `IReadOnlyCollection` instead, to make the required enumeration clearer to the caller? Or is keeping it an `IEnumerable` worth the flexibility? Maybe we could just add a remark to the doc comment that this enumerates `source` before the unzipping operation.